statistics.cpp

```diff
@@ -66,3 +67,91 @@ double Statistics::logLikelihood(const std::valarray<double>& particleSizeData,
+  std::valarray<double> data_counts(1.0, data.size()); // each data point occurred 1 time
+  hist_data.AddToBins(data_counts, data);
+
+  // Step 2 -- Turn distribution into a histogram
+    // Step 2a -- Normalize distribution to create probability mass function (pmf)
+    const double norm_const = distribution.sum();
+    const std::valarray<double> pmf = distribution/norm_const;
+
+    // Step 2b -- Create histogram from pmf
+    Histograms::Histogram hist_ode(hist_prm);
+    hist_ode.AddToBins(pmf, sizes);
+
+  // Step 3 -- Combine histograms to calculate log likelihood
+  double likelihood = 0.0;
+  for (unsigned int bin=0; bin < hist_prm.n_bins; ++bin)
+    {
+      // If the probability is zero and there are no data points, do not contribute anything
+      // If the probability is zero and there are data points, then return most negative number possible
+      // If the probability is non-zero, then calculate normally
+      if (hist_ode.count[bin] == 0 && hist_data.count[bin] == 0)
+        {
+          // do nothing
+        }
+      else if (hist_ode.count[bin] == 0)
+        {
+          // minimum possible value
+          return -std::numeric_limits<double>::max();
+        }
+      else
+        {
+          // normal calculation
+          likelihood += hist_data.count[bin]*std::log(hist_ode.count[bin]);
+        }
+    }
+
+  return likelihood;
+}
+
+
+
+// log likelihood ODE solve integration
+double Statistics::log_likelihood(const std::vector<std::valarray<double>>& data,
+                                  const std::vector<double>& times,
+                                  const Models::ModelsBase& ode_model,
+                                  const Models::ParametersBase& ode_prm,
+                                  const std::valarray<double>& ic,
+                                  const Histograms::Parameters& hist_prm)
+{
+  // Step 1 -- Solve the ODE at each time
+  const std::vector<std::valarray<double>> solutions = Models::integrate_ode_ee_many_times(ic, ode_model, ode_prm, times);
+
+  // Step 2 -- Accumulate log likelihood
+  double likelihood = 0.0;
+  for (unsigned int set_num=0; set_num < data.size(); ++set_num)
+    {
+      // Step 2a -- Extract the particle sizes from the ODE solution
+      const unsigned int smallest = ode_model.getSmallestParticleSize(ode_prm);
+      const unsigned int largest = ode_model.getLargestParticleSize(ode_prm);
+
+      std::valarray<double> sizes(largest - smallest + 1);
+      std::valarray<double> concentration(sizes.size());
+
+      for (unsigned int size = smallest; size < largest+1; ++size)
+        {
+          sizes[size - smallest] = size;
+          concentration[size - smallest] = solutions[set_num][ode_model.particleSizeToIndex(size, ode_prm)];
+        }
+      // Step 2b -- Calculate log likelihood of current data set and add to total
+      likelihood += Statistics::log_likelihood(data[set_num],
+                                               concentration,
```

Comment:
I don't know what's going on with your debugger, but when I run things in my debugger, `concentration` is an array of only zeros. Is that what you expect? In the function called here, I see (where `concentration` is not called `distribution`):
```
    const double norm_const = distribution.sum();
    const std::valarray<double> pmf = distribution/norm_const;
```
So you're dividing by zero. So `concentration` being a zero vector seems to be the bug you're looking for.

Separately, arrays should have plural names.

Reply:
And `concentration` is a zero vector because above you're calling it on the initial data which is zero for all particle sizes.